### PR TITLE
ele/ph mass fix

### PR DESCRIPTION
**Pull request by rgoldouz**
#### PR description:

Hello,
The electron and photon mass are not saved properly and nanoAOD users reported that many times. It is not related to nanoAOD. The problem is available in miniAOD too.
You can check the printout here
http://dalfonso.web.cern.ch/dalfonso/NANO/massDump.txt

I checked and found that the problem is not from the EGamma code but from TLorentzVector root class. I found that the TLorentz vector is set via pxpypzE for all object candidates. It seems fine but when I tested locally for a particle with pxpypzE=(1,1,1,sqrt(3)) I did not get the mass==0. You can easily reproduce it 
root -l
root [1] typedef ROOT::Math::LorentzVector<ROOT::Math::PxPyPzE4D<double> > XYZTLorentzVectorD
root [2] XYZTLorentzVectorD newP4(1,1,1,sqrt(3));
root [3] newP4.M()
(double) -2.1073424e-08

It is because of the precision problem in TLorentz vector and it is recommended to use PxPyPzM to have correct masses for low mass objects.
https://root.cern.ch/doc/master/classROOT_1_1Math_1_1PxPyPzM4D.html 

So I set the electron and photon masses by hand to 0.0005 and 0 GeV , respectively.

#### PR validation:

I generated electron and photon up to reco level using fastsim cmsDriver command and checked the photon and electron mass before and after this fix. Wrong mass values before the fix are corrected after the fix.

please let me know if you have any question.
Thanks ,
Reza

### CODE REVIEW

File: RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc [hunk 1]
```diff
@@ -946,20 +946,25 @@ void GEDPhotonProducer::fillPhotonCollection(edm::Event& evt,
...
       } else if (candidateP4type_ == "fromRefinedSCRegression") {
         newCandidate.setP4(newCandidate.p4(reco::Photon::regression2));
         newCandidate.setCandidateP4type(reco::Photon::regression2);
+        newCandidate.setMass(0);
       }
     } else {
       math::XYZVector gamma_momentum = direction.unit() * scRef->energy();
       math::XYZTLorentzVectorD p4(gamma_momentum.x(), gamma_momentum.y(), gamma_momentum.z(), scRef->energy());
       newCandidate.setP4(p4);
+      newCandidate.setMass(0);
```
Comment (slava77) [inline RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc:967]: use PolarLorentzVector. This is the internal representation of reco candidate p4

Comment (rgoldouz) [reply RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc:967]: Hello Slava,

The setMass function does this transformation.
https://cmssdt.cern.ch/lxr/source/DataFormats/Candidate/interface/ParticleState.h#0159

Do you mean to use,
      math::PtEtaPhiMLorentzVector p4(gamma_momentum.Pt(), gamma_momentum.Eta(), gamma_momentum.Phi(), 0.0);
I could not find Pt function in math::XYZVector.

Please let me know your opinion.
Thanks,
Reza

**Comment by Sam-Harper** [comment]
are we sure we want to set the electron mass to 0.5 MeV? 

As far as I'm aware, all E/gamma code treats the electron as massless and has for ever (modulo this bug where it its not exactly zero due to rounding errors.

Why not continue to set it to zero as was intended.  Is there is a compelling use case to need to adjust the momentum such that the mass is 0.5 MeV? It seems like such a small effect which is best ignored. 

Otherwise, thanks for the fix Reza!

**Comment by rgoldouz** [comment]
Thanks Sam,
I just updated the code with setting ele mass = 0.

Please let me know if you have any further comment.

Thanks,
Reza

File: RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc (hunk 1 above)
Comment (slava77) [reply RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc]: yes, I meant to replace the XYZT p4 construction and setP4;setMass with just a construction of Polar p4 followed by setP4

Comment (rgoldouz) [reply RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc]: Hi Slava,

As I mentioned, math::XYZVector does not have the Pt function. I need to recalculate it and I afraid to introduce bug there. What is wrong with using setMass function?
 It does this transformation automatically using setCartesian function,
 https://cmssdt.cern.ch/lxr/source/DataFormats/Candidate/interface/ParticleState.h#0159

Thanks,
Reza

**Comment by jpata** [comment]
to be a bit more verbose, I think Slava's suggestion to use more precise arithmetic through `typedef math::PtEtaPhiMLorentzVector PolarLorentzVector; ` should be tried here. 
(`mathLLXYZVector` was not suggested and seems to be a miscommunication)

**Comment by rgoldouz** [comment]
Hi @jpata ,
Sorry for my late reply.
As I wrote in my reply to Slava,
If I change the code like this;
math::PtEtaPhiMLorentzVector p4(gamma_momentum.Pt(), gamma_momentum.Eta(), gamma_momentum.Phi(), 0.0);
I need to use gamma_momentum.Pt(), but there is no Pt() function in gamma_momentum which is a math::XYZVector.
I think the corresponding function should be R(). But I afraid to destroy all photon object and found SetMass() function much safer

Please let me know your opinion.

Thanks,
Reza

**Comment by jpata** [comment]
Perhaps something like this?

```C++
math::PtEtaPhiMLorentzVector p4(gamma_momentum.rho(), gamma_momentum.eta(), gamma_momentum.phi(), 0.0);
newCandidate.setP4(p4);
```

**Pull request closed without merging**